thunder/transforms/autodiff.py

```diff
@@ -0,0 +1,301 @@
+                    if nbsym.sym in {prims.get_grad, prims.put_grad} or any(
+                        n in backward_proxy_names for n in argnames
+                    ):
+                        backward_proxy_names.update(p.name for p in nbsym.flat_proxy_outs if p.name not in argnames)
+                        backward_part_bsyms.append(nbsym)
+                    else:
+                        forward_part_bsyms.append(nbsym)
+
+                self.add_processed_bsyms(forward_part_bsyms)
+
+                self.collected_bw_part_bsyms[:0] = backward_part_bsyms
+                return
+
+            # No gradient transform found, need to descend
+            check(not bsym.sym.is_prim, lambda: f"Failed to find a gradient transform for bound symbol {bsym=}")
+
+            # TODO: check if this is needed: the old impl checked whether len(bsym.subsymbols) > 0 except for the special case "torch.nn.functional.dropout" with p=0...
+
+            # OUTPUTS to map
+            self.add_unprocessed_bsyms(bsym.subsymbols[:])
+
+    start_time_ns = time.perf_counter_ns()
+
+    processor = AugmentedForwardProcessor(trace)
+    trace, _ = processor()
+    # trace = thunder.core.transform_common.dce(trace)
+
+    end_time_ns = time.perf_counter_ns()
+    elapsed_time_ns = end_time_ns - start_time_ns
+    elapsed_time_millis = elapsed_time_ns // 1000000
+    trace.set_provenance(
+        thunder.core.trace.TraceProvenance(f"Grad transform pass (took {elapsed_time_millis} milliseconds)")
+    )
+    return trace
+
+
+def split_into_forward_and_backward(joint_trace):
+    forward_part_bsyms = []
+    backward_part_bsyms = []
+
+    return_bsym = joint_trace.bound_symbols[-1]
+    assert return_bsym.sym == prims.python_return
+
+    fw_output = return_bsym.args[0]["output"]
+    assert isinstance(fw_output, tuple)
+
+    grad_outs = [None for _ in fw_output]
+
+    output_pos = {o.name: i for i, o in enumerate(fw_output) if isinstance(o, thunder.TensorProxy)}
+    forward_proxy_names = {o.name for o in thunder.core.pytree.tree_iter(fw_output) if isinstance(o, thunder.Proxy)}
+
+    for bsym in reversed(joint_trace.bound_symbols):
+        if bsym.sym == prims.python_return:
+            continue
+
+        if any((o.name in forward_proxy_names) for o in bsym.flat_proxy_outs):
+            forward_part_bsyms.insert(0, bsym.from_bsym())
+            forward_proxy_names.update(a.name for a in bsym.flat_proxy_args)
+            continue
+
+        if bsym.sym == prims.get_grad:
+            grad_outs[output_pos[bsym.args[0].name]] = bsym.output
+            continue
+
+        backward_part_bsyms.insert(0, bsym.from_bsym())
+
+    # collect needed computation
+    saved_for_backward = {}
+    for bsym in backward_part_bsyms:
+        saved_for_backward.update(
+            (a.name, a)
+            for a in bsym.flat_proxy_args
+            if a.name in forward_proxy_names and a.name not in backward_part_bsyms
```

Comment:
`a.name not in backward_part_bsyms` is always true. Did you mean something else here?

Reply:
Indeed. Thank you for catching.